src/utils.py

```diff
@@ -39,34 +53,44 @@ def remove_invalid_words(self):
 
         if removed_count > 0:
             print(f"Removed {removed_count} invalid words")
-            self.word_list[:] = valid_words
-            self.save_to_file()
+            target_list[:] = valid_words
+            if not self.is_test_mode:
+                self.save_to_file()
         else:
             print("No invalid words found")
 
     def find_scarce_letters(self, num=3):
-        letters_counts = Counter("".join(words.word_list))
+        target_list = self.word_list if self.is_test_mode else words.word_list
```

Comment:
We can discuss later but I think you can safely remove `is_test_mode` for now.

You already handle everything cleanly:
- `self.word_list` = always a copy, so no shared global state = good
- `save_on_change` already disables saving back to file, so that already is your 'test mode' switch.